core/cloudflow-operator/src/main/scala/cloudflow/operator/action/Action.scala

```diff
@@ -305,9 +309,38 @@ final class ProvidedAction[T <: ObjectResource, R <: ObjectResource](
   def executed =
     s"Provided $namespace/$resourceName to next action"
 
-  def execute(client: KubernetesClient)(implicit ec: ExecutionContext, lc: LoggingContext): Future[Action[R]] =
-    for {
-      existing ← client.usingNamespace(namespace).get[T](resourceName)
-      res      <- getAction(existing).execute(client)
-    } yield res
+  def execute(client: KubernetesClient)(implicit sys: ActorSystem, ec: ExecutionContext, lc: LoggingContext): Future[Action[R]] =
+    executeUntil(
+      client,
+      timeout = 1.minute,
+      delayWhenNotFound = 1.second,
+      waiting = 0.second
+    )
+
+  private def executeUntil(
+      client: KubernetesClient,
+      timeout: FiniteDuration,
+      delayWhenNotFound: FiniteDuration,
+      waiting: FiniteDuration
+  )(implicit sys: ActorSystem, ec: ExecutionContext, lc: LoggingContext): Future[Action[R]] =
+    client
+      .usingNamespace(namespace)
+      .getOption[T](resourceName)
+      .flatMap { result =>
+        result
+          .map { existing =>
+            getAction(existing).execute(client)
+          }
+          .getOrElse {
+            if (waiting <= timeout) {
+              after(delayWhenNotFound, sys.scheduler)(executeUntil(client, timeout, delayWhenNotFound, waiting + 1.second))
+            } else {
+              // last attempt expects to be there or fail with error that the resource is not there.
+              client
+                .usingNamespace(namespace)
+                .get[T](resourceName)
+                .flatMap(existing => getAction(existing).execute(client))
+            }
+          }
+      }
```

Comment:
@RayRoestenburg will it be possible to make this retry logic a little bit succinct by using a combinator like https://gist.github.com/viktorklang/9414163 that also implements exponential backoff (like u did) ?

Reply:
Ah right I remember this one now! yeah I just quickly hand-rolled this one ;-) It's a slightly different case, I want to retry when it is not there. If the call somehow is an error it should fail (I think). The retries is more succinct instead of encoding bot the delay, how long waiting, etc. Let me see.

Reply:
yeah I think that version is better, it does not really matter why the resource is not there. Thanks @debasishg !

Reply:
h/t to the Klang.